### DISCUSSION TITLE
addRedactAnnot - With text

### DISCUSSION THREAD

**Discussion by deepanshug** [pymupdf/PyMuPDF#748] [Ideas]
Hi I am trying to use text instead of black rectangle to retain readability context. I have used as below:
"page.addRedactAnnot(area, text = 'sometext', fontname = "Courier", fontsize = 20, fill=(1, 1, 1), text_color=(0,0,0))"

I have a PDF which I am redacting and below is the list of fonts in each page (3 pages):
[(11, 'n/a', 'Type1', 'Courier', 'F1', '')]
[(18, 'n/a', 'Type1', 'Courier', 'F1', '')]
[(25, 'n/a', 'Type1', 'Courier', 'F1', '')]

I am least concerned about the font to be used but I am only intention is to use text instead of black box. E.g. if my document has email address mentioned anywhere, instead of redacting it with black box, I wish to have it replaced with textbox with text "<Email Address>.

appreciate your help in this

**Comment by JorjMcKie** [comment]
Try something like this. I am searching for some word ("pixmap" in this example) and replace all occurrences with the text "enigma".

```python
import fitz

doc = fitz.open("file.pdf")
page = doc[0]
blue = (0, 0, 1)
rl = page.searchFor("pixmap")
for rect in rl:
    fontsize = rect.height / 1.3
    page.addRedactAnnot(
        rect,
        text="enigma",
        text_color=blue,
        fontsize=fontsize,
        align=fitz.TEXT_ALIGN_CENTER,
    )

page.apply_redactions()
doc.save("x.pdf")
```
Before:
![grafik](https://user-images.githubusercontent.com/8290722/101283781-0fe0da80-37b3-11eb-8c8a-504813d4821b.png)

After:

![grafik](https://user-images.githubusercontent.com/8290722/101283821-4d456800-37b3-11eb-9995-0f2dfe7446cc.png)

**Comment by JorjMcKie** [comment]
As you can see, the replacement is not perfectly positioned. This goes back to the fact, that the search algorithm delivers the rectangles with a height equal to the line height. And did not bother to extract the exact insertion point of new text ...
With some effort, better results are achievable - but as a demo this might be sufficient.

**Comment by deepanshug** [comment]
Hi,

First I would like to thank you a providing this library and support.
I tried what you suggested and I was able to see it working.

However you already mentioned about positioning so it did not fit good.
I even tried by reducing area (as you already had suggested on how to reduce area) but it did not work.

However, I am able to move ahead with assignment using this library.

Thanks again for your continuous support.

Thanks
Deepanshu
________________________________

**Comment by JorjMcKie** [comment]
I am working on an improved handling of this.
Under the hood, text insertion for redactions uses `page.insertTextbox`, and I am still making simplifying assumptions in that method as per the primary text insertion point and the line height:
In `page.insertTextbox`, line height is always set to `fontsize * 1.2` and insertion point is always `fontsize` away from the relevant textbox border.
Actually, these values should be font-dependent:
Among a font's properties are "ascender" and "descender" (a negative value). With these values, the correct line height of a font can be computed as `fontsize * (ascender - descender)` and the insertion point distance as `fontsize * ascender`.

I will keep you posted about the progress.

**Comment by deepanshug** [comment]
Thanks Jorj.

I will keep an eye on updates and release notes as well.

Regards
Deepanshu
________________________________

**Comment by JorjMcKie** [comment]
I should have thought of the following possibility earlier: the basic problem is determining the exact insertion point.
This however equals `span["origin"]`, which we can easily determine.
The next thought is to create and **_immediatly apply_** each word occurrence. This empties the corresponding retangle. Then insert the replacement word using the correct insertion point - after adjusting the fontsize of the new text for exact fit:
```python
rl = page.searchFor("pixmap")

for rect in rl:
    for b in page.getText("dict", clip=rect)["blocks"]:
        for l in b["lines"]:
            for span in l["spans"]:
                fsize = span["size"]
                origin = fitz.Point(span["origin"])  # the insertion point
                flags = span["flags"]
                if flags & 2 ** 3:  # is this font monospaced?
                    font = "cour"  # use Courier for new text
                else:
                    font = "helv"  # else stick with Helvetica
    page.addRedactAnnot(rect)  # redact the word
    page.apply_redactions()  # and imediately apply!
    # insert the new text separately - outside redaction

    # First determine length of new text to insert ... this only works
    # for fonts Times-Roman, Helvetica, Courier.
    # There are also ways for arbitrary fonts!
    tl = fitz.getTextlength("enigma", fontname=font, fontsize=fsize)
    # then adjust fontsize, so its fits exactly
    fsize = fsize * rect.width / tl
    page.insertText(origin, "enigma", fontname=font, fontsize=fsize, color=blue)
doc.save("x.pdf")
```
This results in the following:
![grafik](https://user-images.githubusercontent.com/8290722/102065571-1f71ac00-3dcf-11eb-8cc4-23364cdc474d.png)
